_posts/2019-01-14-code-of-the-week.md

```diff
@@ -0,0 +1,127 @@
+module](https://forge.puppet.com/puppet/rsyslog#puppet-rsyslog). There are
+multiple great PRs that we could talk about here. Let us have a look at
+[the latest](https://github.com/voxpupuli/puppet-rsyslog/pull/106).
+
+You can create custom datatypes that are very complex/very strict. They are
+based on multiple scalar types. One very detailed example is this
+(types/actions/outputs/omelasticsearch.pp):
+
+```puppet
+ type Rsyslog::Actions::Outputs::Omelasticsearch = Struct[{
+   server               => Optional[Variant[String[1], Array[String[1]]]],
+   serverport           => Optional[Stdlib::Port],
+   healthchecktimeout   => Optional[Integer],
+   searchindex          => Optional[String[1]],
+   dynsearchindex       => Optional[Enum['on', 'off']],
+   searchtype           => Optional[String[1]],
+   dynsearchtype        => Optional[Enum['on', 'off']],
+   pipelinename         => Optional[String[1]],
+   dynpipelinename      => Optional[Enum['on', 'off']],
+   usehttps             => Optional[Enum['on', 'off']],
+   timeout              => Optional[Pattern[/^([0-9]+)(ms|s|m)$/]],
+   template             => Optional[String[1]],
+   bulkmode             => Optional[Enum['on', 'off']],
+   maxbytes             => Optional[Pattern[/^([0-9]+)[kKmMgGtT]$/]],
+   parent               => Optional[String[1]],
+   dynparent            => Optional[Enum['on', 'off']],
+   uid                  => Optional[String[1]],
+   pwd                  => Optional[String[1]],
+   errorfile            => Optional[Stdlib::Absolutepath],
+   'tls.cacert'         => Optional[Stdlib::Absolutepath],
+   'tls.mycert'         => Optional[Stdlib::Absolutepath],
+   'tls.myprivkey'      => Optional[Stdlib::Absolutepath],
+   bulkid               => Optional[String[1]],
+   dynbulkid            => Optional[Enum['on', 'off']],
+   writeoperation       => Optional[Enum['index', 'create']],
+   retryfailures        => Optional[Enum['on', 'off']],
+   retryruleset         => Optional[String[1]],
+   'ratelimit.interval' => Optional[Integer],
+   'ratelimit.burst'    => Optional[Integer],
+ }]
+```
+
+The rsyslog configuration has so many options, many of them can be nested. In
+the beginning of datatypes in Puppet, we would have used `Array` for the above
+requirement. If we look closely we notice that the custom datatype has a hash
+in the array. So we could update our type from `Array` to `Array[Hash]`. At
+this level the type is already better than the old `validate_array()` function
+and everybody would be happy. But not David! He did not go one, but thousand
+steps further and enforces the complete hash structure. And didn't simply
+create the custom datatype `Rsyslog::Actions::Outputs::Omelasticsearch`,
+he created many more and also tests for them. If we do a quick
+`find types/ -type f -exec cat {} \; | wc -l` we notice that there are 318
+lines of datatype codes. This is amazing! Also
+`find spec/type_aliases/ -type f -exec cat {} \; | wc -l` reveals 649 lines
```

Comment:
aand this is up to 2009 lines of test code.... gruelling, monotonous, boring lines of tests.

Reply:
updated it :D